prov/efa/src/rxr/rxr_pkt_type_misc.c

```diff
@@ -63,45 +63,58 @@ ssize_t rxr_pkt_init_handshake(struct rxr_ep *ep,
 
 	pkt_entry->pkt_size = sizeof(struct rxr_handshake_hdr)
 			      + RXR_NUM_PROTOCOL_VERSION * sizeof(uint64_t);
-	pkt_entry->addr = addr;
+	pkt_entry->addr = tx_entry->addr;
+	pkt_entry->x_entry = tx_entry;
 	return 0;
 }
 
-void rxr_pkt_post_handshake(struct rxr_ep *ep,
-			    struct rxr_peer *peer,
-			    fi_addr_t addr)
+ssize_t rxr_pkt_post_handshake(struct rxr_ep *ep, struct rxr_peer *peer,
+				fi_addr_t addr)
 {
-	struct rxr_pkt_entry *pkt_entry;
+	struct rxr_tx_entry *tx_entry;
 	ssize_t ret;
 
 	assert(!(peer->flags & RXR_PEER_HANDSHAKE_SENT));
 
-	pkt_entry = rxr_pkt_entry_alloc(ep, ep->tx_pkt_efa_pool);
-	if (OFI_UNLIKELY(!pkt_entry))
-		return;
-
-	rxr_pkt_init_handshake(ep, pkt_entry, addr);
-
-	/*
-	 * TODO: Once we start using a core's selective completion capability,
-	 * post the HANDSHAKE packets without FI_COMPLETION.
-	 */
-	ret = rxr_pkt_entry_send(ep, pkt_entry, addr);
+	tx_entry = ofi_buf_alloc(ep->tx_entry_pool);
+	if (OFI_UNLIKELY(!tx_entry)) {
+		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "TX entries exhausted.\n");
+		return -FI_EAGAIN;
+	}
 
-	/*
-	 * Skip sending this handshake on error and try again when processing the
-	 * next REQ from this peer containing the source information
-	 */
+	tx_entry->total_len = 0;
+	tx_entry->addr = addr;
+	tx_entry->msg_id = -1;
+	dlist_init(&tx_entry->queued_pkts);
+	tx_entry->type = RXR_TX_ENTRY;
+	tx_entry->op = ofi_op_msg;
+	tx_entry->state = RXR_TX_REQ;
+
+	tx_entry->send_flags = 0;
+	tx_entry->bytes_acked = 0;
+	tx_entry->bytes_sent = 0;
+	tx_entry->window = 0;
+	tx_entry->rma_iov_count = 0;
+	tx_entry->iov_count = 0;
+	tx_entry->iov_index = 0;
+	tx_entry->iov_mr_start = 0;
+	tx_entry->iov_offset = 0;
+
+#if ENABLE_DEBUG
+	dlist_insert_tail(&tx_entry->tx_entry_entry, &ep->tx_entry_list);
+#endif
+
+	ret = rxr_pkt_post_ctrl_or_queue(ep, RXR_TX_ENTRY, tx_entry, RXR_HANDSHAKE_PKT, 0);
 	if (OFI_UNLIKELY(ret)) {
-		rxr_pkt_entry_release_tx(ep, pkt_entry);
-		if (ret == -FI_EAGAIN)
-			return;
-		FI_WARN(&rxr_prov, FI_LOG_CQ,
-			"Failed to send a HANDSHAKE packet: ret %zd\n", ret);
-		return;
+		// Once error happens, pkt_entry has been released within rxr_pkt_post_ctrl_once
+		rxr_release_tx_entry(ep, tx_entry);
+		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL,
+			"Failed to send a HANDSHAKE packet: %s\n", fi_strerror(-ret));
+		return ret;
 	}
 
 	peer->flags |= RXR_PEER_HANDSHAKE_SENT;
```

Comment:
Consider the possibility of RNR, we need to revisit this logic.
SENT does not mean the packet is received by peer, because it can got RNR. 
I think we need to rename the flag `RXR_PEER_HANDSHAKE_SENT` as `RXR_PEER_HANDSHAKE_SEND_COMPLETE` and set it in `rxr_pkt_handle_handshake_send_completion`

Reply:
Actually, I think even for non-RNR case, it's still not right, as resend could also failure. So we should not set this flag before getting completion.

Reply:
Yes, that is what I meant. Set the flag in `rxr_pkt_handle_handshake_send_completion`, and rename the flag to RXR_PEER_HANDSHAKE_SEND_COMPLETED

Reply:
offline chatted with @zhngaj. This change is not needed, and can cause multiple handshake being sent.

Reply:
Also noted that setting `RXR_PEER_HANDSHAKE_SENT` flag in rxr_pkt_handle_handshake_sent can also cause mutiple handshake being sent, as rxr_pkt_handle_ctrl_sent will not be invoked if rxr_pkt_entry_send returns FI_EAGAIN. 
https://github.com/ofiwg/libfabric/blob/master/prov/efa/src/rxr/rxr_pkt_cmd.c#L339-L348